File: .github/scripts/test-publication-catalog.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "json"
require "pathname"
require "tmpdir"
require_relative "validate-publication-catalog"

BASE_DOCUMENTS = [
  {
    "id" => "overview",
    "source" => "README.md",
    "optional" => false,
    "home" => true
  },
  {
    "id" => "guide",
    "source" => "docs/guide.md",
    "optional" => true,
    "home" => false
  }
].freeze


def deep_copy(value)
  Marshal.load(Marshal.dump(value))
end


def prepare_repository
  Dir.mktmpdir("publication-catalog-test") do |directory|
    root = Pathname.new(directory)
    root.join("docs").mkpath
    root.join("README.md").write("# Overview\n", encoding: "UTF-8")
    root.join("docs/guide.md").write("# Guide\n", encoding: "UTF-8")
    yield root
  end
end


def write_catalog(root, documents: deep_copy(BASE_DOCUMENTS), schema_version: 1, extra: {})
  catalog = {
    "schema_version" => schema_version,
    "documents" => documents
  }.merge(extra)
  path = root.join("docs/publication-catalog.json")
  path.write(JSON.pretty_generate(catalog) + "\n", encoding: "UTF-8")
  path
end

failures = []

prepare_repository do |root|
  begin
    documents = PublicationCatalog.validate(write_catalog(root), root: root)
    expected = %w[overview guide]
    actual = documents.map(&:id)
    failures << "valid catalog: expected #{expected.inspect}, got #{actual.inspect}" unless actual == expected
  rescue StandardError => e
    failures << "valid catalog: unexpected #{e.class}: #{e.message}"
  end
end

invalid_cases = [
  {
    name: "rejects unsupported schema versions",
    pattern: /schema_version must be 1/,
    mutate: lambda do |root|
      write_catalog(root, schema_version: 2)
    end
  },
  {
    name: "rejects unsupported root fields",
    pattern: /unsupported: navigation/,
    mutate: lambda do |root|
      write_catalog(root, extra: { "navigation" => [] })
    end
  },
  {
    name: "rejects duplicate document ids",
    pattern: /Duplicate publication document id: overview/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["id"] = "overview"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects duplicate source paths",
    pattern: /Duplicate publication document source: README\.md/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["source"] = "README.md"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects invalid document ids",
    pattern: /lowercase kebab-case/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["id"] = "Guide_Page"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects unsafe parent traversal",
    pattern: /safe relative POSIX path/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["source"] = "../guide.md"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects backslash paths",
    pattern: /safe relative POSIX path/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["source"] = "docs\\guide.md"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects non-Markdown sources",
    pattern: /must identify a Markdown file/,
    mutate: lambda do |root|
      root.join("docs/guide.txt").write("guide\n", encoding: "UTF-8")
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["source"] = "docs/guide.txt"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects missing source files",
    pattern: /existing regular file/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["source"] = "docs/missing.md"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects symlinked source files",
    pattern: /traverses a symlink/,
    mutate: lambda do |root|
      root.join("docs/guide.md").delete
      File.symlink(root.join("README.md"), root.join("docs/guide.md"))
      write_catalog(root)
    end
  },
  {
    name: "rejects catalogs without a home document",
    pattern: /exactly one home document/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents.each { |document| document["home"] = false }
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects multiple home documents",
    pattern: /exactly one home document/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents.each { |document| document["home"] = true }
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects an optional home document",
    pattern: /home document must not be optional/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[0]["optional"] = true
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects non-boolean optional values",
    pattern: /optional must be boolean/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1]["optional"] = "yes"
      write_catalog(root, documents: documents)
    end
  },
  {
    name: "rejects missing document fields",
    pattern: /missing: home/,
    mutate: lambda do |root|
      documents = deep_copy(BASE_DOCUMENTS)
      documents[1].delete("home")
      write_catalog(root, documents: documents)
    end
  }
]

invalid_cases.each do |test_case|
  prepare_repository do |root|
    catalog_path = test_case.fetch(:mutate).call(root)
    begin
      PublicationCatalog.validate(catalog_path, root: root)
      failures << "#{test_case.fetch(:name)}: validation unexpectedly succeeded"
    rescue PublicationCatalog::ValidationError => e
      unless test_case.fetch(:pattern).match?(e.message)
        failures << "#{test_case.fetch(:name)}: unexpected diagnostic #{e.message.inspect}"
      end
    rescue StandardError => e
      failures << "#{test_case.fetch(:name)}: unexpected #{e.class}: #{e.message}"
    end
  end
end

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

puts "Publication catalog tests passed (#{invalid_cases.length + 1} cases)."

File: .github/scripts/test_publication_catalog.py
#!/usr/bin/env python3
"""Regression tests for the publication catalog validator."""

from __future__ import annotations

import copy
import json
import re
import sys
import tempfile
from pathlib import Path

from validate_publication_catalog import ValidationError, validate


BASE_DOCUMENTS = [
    {"id": "overview", "source": "README.md", "optional": False, "home": True},
    {"id": "guide", "source": "docs/guide.md", "optional": True, "home": False},
]


def prepare_repository() -> tempfile.TemporaryDirectory[str]:
    temporary = tempfile.TemporaryDirectory(prefix="publication-catalog-test-")
    root = Path(temporary.name)
    (root / "docs").mkdir()
    (root / "README.md").write_text("# Overview\n", encoding="utf-8")
    (root / "docs/guide.md").write_text("# Guide\n", encoding="utf-8")
    return temporary


def write_catalog(
    root: Path,
    *,
    documents: list[dict[str, object]] | None = None,
    schema_version: object = 1,
    extra: dict[str, object] | None = None,
) -> Path:
    catalog = {
        "schema_version": schema_version,
        "documents": copy.deepcopy(BASE_DOCUMENTS if documents is None else documents),
    }
    if extra:
        catalog.update(extra)
    path = root / "docs/publication-catalog.json"
    path.write_text(json.dumps(catalog, indent=2) + "\n", encoding="utf-8")
    return path


def _mutated_catalog(root: Path, index: int, field: str, value: object) -> Path:
    documents = copy.deepcopy(BASE_DOCUMENTS)
    documents[index][field] = value
    return write_catalog(root, documents=documents)


def _write_invalid_utf8(root: Path) -> Path:
    path = write_catalog(root)
    path.write_bytes(path.read_bytes().replace(b"guide", b"guide\xff", 1))
    return path


def _write_non_markdown(root: Path) -> Path:
    (root / "docs/guide.txt").write_text("guide\n", encoding="utf-8")
    return _mutated_catalog(root, 1, "source", "docs/guide.txt")


def _write_symlinked_source(root: Path) -> Path:
    (root / "docs/guide.md").unlink()
    (root / "docs/guide.md").symlink_to(root / "README.md")
    return write_catalog(root)


def _set_all_home(root: Path, value: bool) -> Path:
    documents = copy.deepcopy(BASE_DOCUMENTS)
    for document in documents:
        document["home"] = value
    return write_catalog(root, documents=documents)


def _write_missing_home(root: Path) -> Path:
    documents = copy.deepcopy(BASE_DOCUMENTS)
    del documents[1]["home"]
    return write_catalog(root, documents=documents)


def run() -> int:
    failures: list[str] = []

    with prepare_repository() as directory:
        root = Path(directory)
        try:
            documents = validate(write_catalog(root), root=root)
            actual = [document.id for document in documents]
            expected = ["overview", "guide"]
            if actual != expected:
                failures.append(f"valid catalog: expected {expected!r}, got {actual!r}")
        except Exception as exc:  # noqa: BLE001 - record all harness failures.
            failures.append(f"valid catalog: unexpected {type(exc).__name__}: {exc}")

    invalid_cases = [
        (
            "rejects unsupported schema versions",
            r"schema_version must be integer 1 or 3",
            lambda root: write_catalog(root, schema_version=2),
        ),
        (
            "rejects unsupported root fields",
            r"unsupported: navigation",
            lambda root: write_catalog(root, extra={"navigation": []}),
        ),
        (
            "rejects malformed UTF-8 before JSON parsing",
            r"content is not valid UTF-8",
            _write_invalid_utf8,
        ),
        (
            "rejects duplicate document ids",
            r"Duplicate publication document id: overview",
            lambda root: _mutated_catalog(root, 1, "id", "overview"),
        ),
        (
            "rejects duplicate source paths",
            r"Duplicate publication document source: README\.md",
            lambda root: _mutated_catalog(root, 1, "source", "README.md"),
        ),
        (
            "rejects invalid document ids",
            r"lowercase kebab-case",
            lambda root: _mutated_catalog(root, 1, "id", "Guide_Page"),
        ),
        (
            "rejects unsafe parent traversal",
            r"safe relative POSIX path",
            lambda root: _mutated_catalog(root, 1, "source", "../guide.md"),
        ),
        (
            "rejects backslash paths",
            r"safe relative POSIX path",
            lambda root: _mutated_catalog(root, 1, "source", "docs\\guide.md"),
        ),
        (
            "rejects non-Markdown sources",
            r"must identify a Markdown file",
            _write_non_markdown,
        ),
        (
            "rejects missing source files",
            r"existing regular file",
            lambda root: _mutated_catalog(root, 1, "source", "docs/missing.md"),
        ),
        (
            "rejects symlinked source files",
            r"traverses a symlink",
            _write_symlinked_source,
        ),
        (
            "rejects catalogs without a home document",
            r"exactly one home document",
            lambda root: _set_all_home(root, False),
        ),
        (
            "rejects multiple home documents",
            r"exactly one home document",
            lambda root: _set_all_home(root, True),
        ),
        (
            "rejects an optional home document",
            r"home document must not be optional",
            lambda root: _mutated_catalog(root, 0, "optional", True),
        ),
        (
            "rejects non-boolean optional values",
            r"optional must be boolean",
            lambda root: _mutated_catalog(root, 1, "optional", "yes"),
        ),
        (
            "rejects missing document fields",
            r"missing: home",
            _write_missing_home,
        ),
    ]

    for name, pattern, mutate in invalid_cases:
        with prepare_repository() as directory:
            root = Path(directory)
            catalog_path = mutate(root)
            try:
                validate(catalog_path, root=root)
                failures.append(f"{name}: validation unexpectedly succeeded")
            except ValidationError as exc:
                if re.search(pattern, str(exc)) is None:
                    failures.append(f"{name}: unexpected diagnostic {str(exc)!r}")
            except Exception as exc:  # noqa: BLE001 - record all harness failures.
                failures.append(f"{name}: unexpected {type(exc).__name__}: {exc}")

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1

    print(f"Publication catalog tests passed ({len(invalid_cases) + 1} cases).")
    return 0


if __name__ == "__main__":
    raise SystemExit(run())
